packages/run-protocol/src/runStake/runStakeManager.js

```diff
@@ -6,123 +6,165 @@ import { makeRatio } from '@agoric/zoe/src/contractSupport/ratio.js';
  * @param {Object} timing
  * @param {ERef<TimerService>} timing.timerService
  * @param {bigint} timing.chargingPeriod
  * @param {bigint} timing.recordingPeriod
  * @param {bigint} timing.startTimeStamp
  *
- * @typedef {ReturnType<typeof makeRunStakeManager>} RunStakeManager
+ * @returns {State}
  */
-export const makeRunStakeManager = (
+const initState = (
   zcf,
   debtMint,
   brands,
-  paramManager,
-  mintAndReallocateWithFee,
-  burnDebt,
-  { timerService, chargingPeriod, recordingPeriod, startTimeStamp },
+  mintPowers,
+  { chargingPeriod, recordingPeriod, startTimeStamp, timerService },
 ) => {
-  /** @param { Amount<'copyBag'>} attestationGiven */
-  const maxDebtForLien = attestationGiven => {
-    const mintingRatio = paramManager.getMintingRatio();
-    assert.equal(
-      mintingRatio.numerator.brand,
-      brands.debt,
-      X`${mintingRatio} not in Debt / Stake`,
-    );
-    assert.equal(
-      mintingRatio.denominator.brand,
-      brands.Stake,
-      X`${mintingRatio} not in Debt / Stake`,
-    );
-    assert.equal(
-      attestationGiven.brand,
-      brands.Attestation,
-      X`Invalid Attestation ${attestationGiven}. Expected brand ${brands.Attestation}`,
-    );
-    fit(attestationGiven.value, M.bagOf([M.string(), M.bigint()]));
-    const [[_addr, valueLiened]] = getCopyBagEntries(attestationGiven.value);
-    const amountLiened = AmountMath.make(brands.Stake, valueLiened);
-    const maxDebt = floorMultiplyBy(amountLiened, mintingRatio);
-    return { maxDebt, amountLiened };
-  };
+  const totalDebt = AmountMath.makeEmpty(brands.debt, 'nat');
+  const compoundedInterest = makeRatio(100n, brands.debt); // starts at 1.0, no interest
+  const latestInterestUpdate = startTimeStamp;
 
-  let totalDebt = AmountMath.makeEmpty(brands.debt, 'nat');
-  let compoundedInterest = makeRatio(100n, brands.debt); // starts at 1.0, no interest
-  let latestInterestUpdate = startTimeStamp;
-  const { updater: assetUpdater, notifier: assetNotifer } = makeNotifierKit(
+  const { updater: assetUpdater, notifier: assetNotifier } = makeNotifierKit(
     harden({
       compoundedInterest,
-      interestRate: paramManager.getInterestRate(),
+      interestRate: mintPowers.getGovernedParams().getInterestRate(),
       latestInterestUpdate,
       totalDebt,
     }),
   );
 
+  // ??? does this promise need to be a presence?
+  const periodNotifier = E(timerService).makeNotifier(0n, recordingPeriod);
+  const { zcfSeat: poolIncrementSeat } = zcf.makeEmptySeatKit();
+
+  return {
+    assetNotifier,
+    assetUpdater,
+    brands,
+    chargingPeriod,
+    compoundedInterest,
+    debtMint,
+    latestInterestUpdate,
+    mintPowers,
+    periodNotifier,
+    poolIncrementSeat,
+    recordingPeriod,
+    startTimeStamp,
+    totalDebt,
+    zcf,
+  };
+};
+
+/**
+ *
+ * @param {MethodContext} context
+ */
+const finish = ({ state }) => {
+  const { debtMint, mintPowers, periodNotifier, poolIncrementSeat, zcf } =
```

Comment:
This use of `finish` (and accessing `state` within it) is beyond our intentions/expectations for virtual objects. Using `state` from outside behavior functions feels like a violation of the intended encapsulation.

I think a better (or less surprising, at least to me) way to approach this would be use `defineKindMulti` to create two facets: one with all the main behavior, and a second with just this `chargeAllVaults` behavior. Then `finish` could wire up the `observeNotifier` to the second facet (ideally just give it behavior methods like `updateState` and `fail` and `finish`).

I don't know how `observeNotifier` is currently implemented, but if we can eventually change it to use a DurableMapStore, then it could store the vref of the second facet, and then it wouldn't even need to keep a Representative in RAM. That doesn't matter here because the runStakeManager is probably a singleton object, but it's probably not the worst pattern to stick to.

Using multiple facets would require an extra wrapper function at the end, so the external caller gets the main facet but never the second one (which is only used for the `observeNotifier`).

With that, the `periodNotifier` wouldn't need to be created during `init` or stored in `state`. It would be created by the wrapper function, which would do the wiring. The `timerService` wouldn't be passed into `init` either. `init` would only get the things that must go into `state`, and `state` would only hold the things needed by behavior methods.

We added `finish` to enable registration of facets to happen before a caller got to see them, but since it doesn't let you influence the return value of the generated construction function (`makeRunStakeManager`), specifically allowing you to remove internal facets from the return value, it only really provides half the job.

I kind of want to get rid of `finish`, or at least remove `state` from its arguments (only passing the facets). I think it was more important in an earlier design for virtual objects (@FUDCo might remember more here), where there was a certain critical moment when object identities had been established but certain code couldn't do anything with them yet. If its current utility is a per-instance post-facet-construction registration/cross-wiring function, but you need an external wrapper anyways (to limit the facets being returned), then it might be just a tempting nuisance, and maybe we should remove it. Or, go all-in and make it capable of doing both jobs (but that would either make it mandatory, or would do something like make the return value mandatory, which I think would be error prone).

@mhofman has some ideas for a different API that would allow this notifier-registration stuff to happen during `init`, which would also remove `finish`, but that feels like a signficant API change and I'm afraid of those at this stage :).

Reply:
Thanks for explaining all that about the API design intent, better practices now, and how things might change.

Given the fluctuations in the API, my bar right now is _what works_. I commit to fixing anything that breaks as the API changes (including tighter constraints on how it's used). I expect there will be new hurdles to make the vobjs durable, but I prefer to jump those when I get to them. These big refactors tend to have high churn in merge conflicts so I want this branch to live as briefly as possible.

With [3828638](https://github.com/Agoric/agoric-sdk/pull/5135/commits/3828638d233c88e9882823278369ffef7c3c1e97) and [bcfa3f2](https://github.com/Agoric/agoric-sdk/pull/5135/commits/bcfa3f2d4d6c75f1a9446d2d8c1f0901c7e22dd6) does this PR seem an acceptable use of the vobj API?